### PR TITLE
Hamiltonian cycles

### PR DESCRIPTION
**Pull request by le27**
Implements a new property `IsHamiltonianDigraph` which checks if a given digraph is hamiltonian, and a new attribute `HamiltonianPath` which finds a hamiltonian path if there is one.

### CODE REVIEW

**Comment by wilfwilson** [comment]
@le27 Please rebase this on top of the most recent commit in the `master` branch - doing so will require you to resolve one or more conflicts. This way you'll get the most up-to-date Travis setup, so that your pull request will be tested properly.

**Comment by james-d-mitchell** [comment]
@le27 are you going to update this at some point?

**Comment by le27** [comment]
sorry I thought I had fixed this a while ago I'll fix it now.

**Comment by james-d-mitchell** [comment]
So, your second commit says that you are implementing a more efficient version of this, but in the test files, only very small graphs are tested. Can you add some tests (probably in the `extreme` directory) that actually test the performance of your implementation?

File: doc/prop.xml [hunk 1]
```diff
@@ -788,3 +788,38 @@ true
 </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="IsHamiltonianDigraph">
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If <A>digraph</A> is Hamiltonian, this property returns
+    <K>true</K>.<P/>
```
Comment (james-d-mitchell) [inline doc/prop.xml]: So what does it return if it is not Hamiltonian?

File: doc/prop.xml [hunk 2]
```diff
@@ -788,3 +788,38 @@ true
...
+
+    A Hamiltonian digraph is a digraph which has a circuit visiting 
+    each vertex exactly once.
+    Note the empty digraphs on 0 and 1 vertices are considered
```
Comment (james-d-mitchell) [inline doc/prop.xml]: Do you mean digraphs with no edges? Or do you just mean "any digraph with 0 or 1 vertices is Hamiltonian."? I note in the code you really be a digraphs with 0 or 1 vertices and no edges, so a digraph with an 1 vertex and a loop is not Hamiltonian?

File: doc/prop.xml [hunk 3]
```diff
@@ -788,3 +788,38 @@ true
...
 
+<#GAPDoc Label="IsHamiltonianDigraph">
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If <A>digraph</A> is Hamiltonian, this property returns
+    <K>true</K>.<P/>
+
+    A Hamiltonian digraph is a digraph which has a circuit visiting
```
Comment (james-d-mitchell) [inline doc/prop.xml]: "<E>Hamiltonian</E>" and wouldn't it be better to write, something like "A digraph is <E>Hamiltonian</E>..."?

Comment (james-d-mitchell) [inline doc/prop.xml]: Probably you should also elaborate the meaning of circuit in this context too.

File: doc/prop.xml (hunk 2 above)
Comment (james-d-mitchell) [inline doc/prop.xml]: You should also say something here about the complexity of the implemented algorithm.

File: gap/prop.gi [hunk 4]
```diff
@@ -555,3 +555,108 @@ function(gr)
...
+InstallMethod(IsHamiltonianDigraph, "for a digraph",
+[IsDigraph],
+function(g)
+  local g2, out, Indegs, Outdegs, Degrees, n, check,
+  dominatedcheck, dominatingordominatedcheck, i, j, k, dp;
+
+  if DigraphNrVertices(g) <= 1 and IsEmptyDigraph(g) then
+     return true;
+  fi;
+  if not IsStronglyConnectedDigraph(g) then
```
Comment (james-d-mitchell) [inline gap/prop.gi]: `if` -> `elif` and remove the previous line

File: gap/prop.gi [hunk 5]
```diff
@@ -555,3 +555,108 @@ function(gr)
   return true;
 
 end);
+
+#
+
```
Comment (james-d-mitchell) [inline gap/prop.gi]: There are no comments in your code here. You should add either a description of the algorithm you are using, or a link/url/reference for the underlying algorithm. Without that this will be harder to maintain this code.

File: gap/prop.gi [hunk 6]
```diff
@@ -555,3 +555,108 @@ function(gr)
...
+  local g2, out, Indegs, Outdegs, Degrees, n, check,
+  dominatedcheck, dominatingordominatedcheck, i, j, k, dp;
+
+  if DigraphNrVertices(g) <= 1 and IsEmptyDigraph(g) then
+     return true;
+  fi;
+  if not IsStronglyConnectedDigraph(g) then
+     return false;
+  fi;
+  g2 := DigraphRemoveLoops(g);
```
Comment (james-d-mitchell) [inline gap/prop.gi]: This is unnecessarily wasteful, since the digraph `g` is copied here even if it has no loops anyway. Better do `if DigraphHasLoops(g) then ...` and otherwise set `g2 := g`

**Comment by wilfwilson** [comment]
@james-d-mitchell is this the sort of thing you were on about, for the default DFS?

```
is_hamiltonian := function(gr)
  local iter, n;
  iter := IteratorOfPaths(gr, 1, 1);
  n := DigraphNrVertices(gr) + 1;
  while not IsDoneIterator(iter) do
    if Length(NextIterator(iter)[1]) = n then
      return true;
    fi;
  od;
  return false;
end;
```
We iterate over paths from the first vertex to itself and return `true` if we find one that contains every vertex.

**Comment by wilfwilson** [comment]
@james-d-mitchell Here's an example where the above method works

```
gap> gr := DigraphFromGraph6String("SXqk_CiF?c?O?WACAwNIGGgpijPyB_LM?");
```

**Comment by james-d-mitchell** [comment]
@wilfwilson yes something like this.

**Comment by james-d-mitchell** [comment]
You can also use `DigraphMonomorphism(CycleDigraph(NrVertices(gr)), gr) <> fail;`. I just discussed this with Luke and he is going to experiment to decide which method to use in which situation.

File: doc/oper.xml [hunk 7]
```diff
@@ -1503,3 +1503,41 @@ true
   </Description>
 </ManSection>
 <#/GAPDoc>
+  
+<#GAPDoc Label="HamiltonianPath">
+<ManSection>
+  <Prop Name="HamiltonianPath" Arg="digraph"/>
+  <Returns>A List or <K>fail</K>.</Returns>
```
Comment (james-d-mitchell) [inline doc/oper.xml]: "List" -> "list" (no capital)

File: doc/oper.xml [hunk 8]
```diff
@@ -1503,3 +1503,41 @@ true
...
+  <Description>
+    Returns a <E>Hamiltonian</E> path if one exists, <K>fail</K> if not.
```
Comment (james-d-mitchell) [inline doc/oper.xml]: Only emphasis words when you are defining them, here you are not defining this, so lose the emphasis.

File: doc/oper.xml [hunk 9]
```diff
@@ -1503,3 +1503,41 @@ true
...
+
+    A <E>Hamiltonian</E> path is a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
+    Note the empty digraphs on 0 and 1 vertices are considered
+    Hamiltonian.<P/>
+
+    The method used in this operation has worst case complexity <M>O(n!)</M>
```
Comment (james-d-mitchell) [inline doc/oper.xml]: I am not sure that this is correct, since you no longer use the naive DFS approach. Maybe remove this sentence altogether.

Comment (james-d-mitchell) [inline doc/oper.xml]: Or say something like "This method uses DigraphMonomorphism, and hence has worst case complexity the same as DigraphMonomorphism."

File: doc/prop.xml [hunk 10]
```diff
@@ -19,7 +19,7 @@
       Oper="CycleDigraph"/>.<P/>
 
     A digraph is a <E>cycle</E> if and only if it is strongly connected and has
-    the same number of edges as vertices. 
+    the same number of edges as vertices.
```
Comment (james-d-mitchell) [inline doc/prop.xml]: This change is not related to this PR, so please drop it.

File: doc/prop.xml [hunk 11]
```diff
@@ -46,7 +46,7 @@ true]]></Example>
   <Returns><K>true</K> or <K>false</K>.</Returns>
   <Description>
     A connected digraph is <E>biconnected</E> if it is still connected (in the
-    sense of <Ref Prop="IsConnectedDigraph"/>) when any vertex is removed. 
+    sense of <Ref Prop="IsConnectedDigraph"/>) when any vertex is removed.
     <C>IsBiconnectedDigraph</C> returns <K>true</K> if the digraph
```
Comment (james-d-mitchell) [inline doc/prop.xml]: Same again, please drop this unrelated change.

File: doc/prop.xml [hunk 12]
```diff
@@ -79,7 +79,7 @@ false]]></Example>
     <Ref Prop="IsAntisymmetricDigraph"/> and <Ref Prop="IsTransitiveDigraph"/>.
     For a partial order digraph <A>digraph</A>, the corresponding partial order
     is the relation <M>\leq</M>, defined by <M>x \leq y</M> if and only if
-    <C>[x, y]</C> is an edge of <A>digraph</A>.    
+    <C>[x, y]</C> is an edge of <A>digraph</A>.
```
Comment (james-d-mitchell) [inline doc/prop.xml]: Same again, please drop this unrelated change.

File: doc/prop.xml [hunk 13]
```diff
@@ -106,7 +106,7 @@ true
   <Prop Name="IsLatticeDigraph" Arg="digraph"/>
   <Returns><K>true</K> or <K>false</K>.</Returns>
   <Description>
-    
+
     <C>IsMeetSemilatticeDigraph</C> returns <K>true</K> if the digraph
```
Comment (james-d-mitchell) [inline doc/prop.xml]: Same again, please drop this unrelated change.

File: doc/prop.xml [hunk 14]
```diff
@@ -850,3 +850,41 @@ true
 </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="IsHamiltonianDigraph">
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If <A>digraph</A> is <E>Hamiltonian</E>, this property returns
```
Comment (james-d-mitchell) [inline doc/prop.xml]: No emphasis.

File: doc/prop.xml [hunk 15]
```diff
@@ -850,3 +850,41 @@ true
...
+    <K>true</K>, and <K>false</K> if it is not. <P/>
+
+    A digraph is <E>Hamiltonian</E> if it has a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
+    Note the empty digraphs on 0 and 1 vertices are considered
+    Hamiltonian.<P/>
+
+    The method used in this operation has worst case complexity <M>O(n!)</M>
```
Comment (james-d-mitchell) [inline doc/prop.xml]: Same comment as before.

File: doc/z-chap5.xml [hunk 16]
```diff
@@ -33,7 +33,13 @@
     <#Include Label="IsDirectedTree">
     <#Include Label="IsUndirectedTree">
     <#Include Label="IsEulerianDigraph">
+    <#Include Label="IsHamiltonianDigraph">
+    <#Include Label="HamiltonianPath">
```
Comment (james-d-mitchell) [inline doc/z-chap5.xml]: This is in the wrong part of the manual, it should be where the other attributes are, not in the section about properties.

File: doc/z-chap5.xml [hunk 17]
```diff
@@ -33,7 +33,13 @@
...
     <#Include Label="IsCycleDigraph">
   </Section>
-  
+
+  <Section><Heading>Partial Orders</Heading>
+    <#Include Label="IsPartialOrderDigraph">
+    <#Include Label="IsMeetSemilatticeDigraph">
+  </Section>
```
Comment (james-d-mitchell) [inline doc/z-chap5.xml]: This section was deleted, you need to rebase on the current HEAD of the master branch.

File: gap/oper.gd [hunk 18]
```diff
@@ -98,6 +98,8 @@ DeclareOperation("PartialOrderDigraphMeetOfVertices",
 [IsDigraph, IsPosInt, IsPosInt]);
 DeclareOperation("DigraphClosure", [IsDigraph, IsPosInt]);
 
+DeclareOperation("HamiltonianPath", [IsDigraph]);
```
Comment (james-d-mitchell) [inline gap/oper.gd]: This should be declared as an attribute not an operation, so that it is not recalculated every time this is called. It should therefore go in the `attr.g*` files.

File: gap/oper.gi [hunk 19]
```diff
@@ -1841,3 +1841,38 @@ function(gr, edges)
...
+      return [];
+    else
+      return [1];
+    fi;
+  elif not IsStronglyConnectedDigraph(gr) then
+    return fail;
+  fi;
+
+  out := OutNeighbours(gr);
+  obviouspath := true;
```
Comment (james-d-mitchell) [inline gap/oper.gi]: Is it worth doing this? I think it probably isn't. Since `DigraphMonomorphism` will do this anyway.

File: gap/oper.gi [hunk 20]
```diff
@@ -1841,3 +1841,38 @@ function(gr, edges)
...
+    if not i + 1 in out[i] then
+      obviouspath := false;
+      break;
+    fi;
+  od;
+  if obviouspath and 1 in out[DigraphNrVertices(gr)] then
+    return [1 .. DigraphNrVertices(gr)] + 0;
+  fi;
+
+  map := DigraphMonomorphism(CycleDigraph(DigraphNrVertices(gr)), gr);
```
Comment (james-d-mitchell) [inline gap/oper.gi]: This only works if the number of vertices in the digraph is at most 512, so you should check for this and use the method suggested by Wilf if the number of vertices is greater than 512.

File: gap/prop.gi [hunk 21]
```diff
@@ -567,3 +567,109 @@ function(gr)
...
+  fi;
+
+  if IsMultiDigraph(gr) then
+    gr := DigraphRemoveAllMultipleEdges(gr);
+  fi;
+  if DigraphHasLoops(gr) then
+    gr := DigraphRemoveLoops(gr);
+  fi;
+
+  if IsCycleDigraph(gr) then
```
Comment (james-d-mitchell) [inline gap/prop.gi]: I don't know if this check is worth it either.

File: tst/standard/oper.tst [hunk 22]
```diff
@@ -1993,6 +1993,123 @@ gap> edges := [[1, 1], [2, 3]];
 gap> IsMaximalMatching(gr, edges);
 true
```
Comment (james-d-mitchell) [inline tst/standard/oper.tst]: These tests are in the wrong file.

File: doc/oper.xml [hunk 23]
```diff
@@ -1503,3 +1503,41 @@ true
   </Description>
 </ManSection>
 <#/GAPDoc>
+  
+<#GAPDoc Label="HamiltonianPath">
+<ManSection>
+  <Prop Name="HamiltonianPath" Arg="digraph"/>
+  <Returns>A list or <K>fail</K>.</Returns>
+  <Description>
+    Returns a <E>Hamiltonian</E> path if one exists, <K>fail</K> if not.
```
Comment (james-d-mitchell) [inline doc/oper.xml]: No emphasis here

File: doc/oper.xml [hunk 24]
```diff
@@ -1503,3 +1503,41 @@ true
...
+
+    A Hamiltonian path is a directed circuit visiting
```
Comment (james-d-mitchell) [inline doc/oper.xml]: Emphasis here.

File: doc/oper.xml [hunk 25]
```diff
@@ -1503,3 +1503,41 @@ true
...
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
+    Note the empty digraphs on 0 and 1 vertices are considered
+    Hamiltonian.<P/>
+
+    The method used in this operation has the worst case complexity as
+    DigraphMonomorphism.
```
Comment (james-d-mitchell) [inline doc/oper.xml]: \<Ref Oper="DigraphMonomorphism"/\>

File: doc/prop.xml [hunk 26]
```diff
@@ -850,3 +850,41 @@ true
 </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="IsHamiltonianDigraph">
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If digraph is <E>Hamiltonian</E>, this property returns
```
Comment (james-d-mitchell) [inline doc/prop.xml]: No emphasis.

File: doc/prop.xml [hunk 27]
```diff
@@ -850,3 +850,41 @@ true
...
+    <K>true</K>, and <K>false</K> if it is not. <P/>
+
+    A digraph is <E>Hamiltonian</E> if it has a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
+    Note the empty digraphs on 0 and 1 vertices are considered
+    Hamiltonian.<P/>
+
+    The method used in this operation has the worst case complexity as
+    DigraphMonomorphism.
```
Comment (james-d-mitchell) [inline doc/prop.xml]: Reference

File: gap/attr.gd [hunk 28]
```diff
@@ -69,6 +69,7 @@ DeclareOperation("DIGRAPHS_MaximalSymmetricSubdigraph", [IsDigraph, IsBool]);
 
 DeclareAttribute("UndirectedSpanningTree", IsDigraph);
 DeclareAttribute("UndirectedSpanningForest", IsDigraph);
+DeclareOperation("HamiltonianPath", [IsDigraph]);
```
Comment (wilfwilson) [inline gap/attr.gd]: This should be an attribute in my opinion. Since its value can't change.

File: doc/attr.xml [hunk 29]
```diff
@@ -1310,3 +1310,42 @@ fail]]></Example>
   </Description>
 </ManSection>
 <#/GAPDoc>
+  
+<#GAPDoc Label="HamiltonianPath">
+<ManSection>
+  <Prop Name="HamiltonianPath" Arg="digraph"/>
```
Comment (wilfwilson) [inline doc/attr.xml]: This shouldn't be a `Prop` - the type should agree with how `HamiltonianPath` (currently it's an Operation, and so it should say `Oper` here, although further down I suggest that `HamiltonianPath` should instead be declared as an attribute, and so this should be `Attr`).

File: doc/attr.xml [hunk 30]
```diff
@@ -1310,3 +1310,42 @@ fail]]></Example>
...
+  <Returns>A list or <K>fail</K>.</Returns>
+  <Description>
+    Returns a Hamiltonian path if one exists, <K>fail</K> if not.
+
+    A <E>Hamiltonian path</E> is a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
```
Comment (wilfwilson) [inline doc/attr.xml]: You should actually read these definitions, since currently this documentation is not quite true. We define a directed circuit to be a particular kind sequence of vertices and edges. Your method just returns a sequence of vertices, so it's *not* returning a directed circuit, as you state. Please look at the documentation for `DigraphAllSimpleCircuits` to see an example of how to be more precise about this. Also - you should say `cycle` rather than `circuit` - again, look at the definition.

You could say something like `A <E>Hamiltonian path</E> of a digraph with <C>n</C> vertices is directed cycle of length <C>n</C>. If <A>digraph</A> is a digraph that contains a Hamiltonian path, then this function returns one, described in the form used by <Ref Attr="DigraphAllSimpleCircuits"/>...` and whatever else you need to make this bit of documentation correct and consistent.

File: doc/attr.xml [hunk 31]
```diff
@@ -1310,3 +1310,42 @@ fail]]></Example>
...
+    Note the empty digraphs on 0 and 1 vertices are considered
+    Hamiltonian.<P/>
+
+    The method used in this operation has the worst case complexity as
```
Comment (wilfwilson) [inline doc/attr.xml]: It's not currently an `operation` - so make sure this agrees with however it is declared.

File: doc/attr.xml [hunk 32]
```diff
@@ -1310,3 +1310,42 @@ fail]]></Example>
...
+  <Prop Name="HamiltonianPath" Arg="digraph"/>
+  <Returns>A list or <K>fail</K>.</Returns>
+  <Description>
+    Returns a Hamiltonian path if one exists, <K>fail</K> if not.
+
+    A <E>Hamiltonian path</E> is a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
+    Note the empty digraphs on 0 and 1 vertices are considered
+    Hamiltonian.<P/>
```
Comment (wilfwilson) [inline doc/attr.xml]: You've not defined what Hamiltonian is so this doesn't make sense on its own. Instead, you could say something like `If <A>digraph</A> has 0 or 1 vertices, then <C>HamiltonianPath</C> returns [] or [1], respectively.`

File: doc/attr.xml [hunk 33]
```diff
@@ -1310,3 +1310,42 @@ fail]]></Example>
...
+
+    The method used in this operation has the worst case complexity as
+    DigraphMonomorphism. See <Ref Oper="DigraphMonomorphism"/> for information
```
Comment (wilfwilson) [inline doc/attr.xml]: just have `...complexity as <Ref Oper="DigraphMonomorphism"/>`.

File: doc/prop.xml [hunk 34]
```diff
@@ -852,3 +852,42 @@ true
 </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="IsHamiltonianDigraph">
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If digraph is Hamiltonian, this property returns
```
Comment (wilfwilson) [inline doc/prop.xml]: `A` tags around `digraph`, since it's the argument. Also have the word `then` after the comma: `, then this...`

File: doc/prop.xml [hunk 35]
```diff
@@ -852,3 +852,42 @@ true
...
+    <K>true</K>, and <K>false</K> if it is not. <P/>
+
+    A digraph is <E>Hamiltonian</E> if it has a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
+    Note the empty digraphs on 0 and 1 vertices are considered
```
Comment (wilfwilson) [inline doc/prop.xml]: Add "to be", i.e. `considered to be Hamiltonian`

File: doc/prop.xml [hunk 36]
```diff
@@ -852,3 +852,42 @@ true
...
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If digraph is Hamiltonian, this property returns
+    <K>true</K>, and <K>false</K> if it is not. <P/>
+
+    A digraph is <E>Hamiltonian</E> if it has a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
```
Comment (wilfwilson) [inline doc/prop.xml]: directed circuit -> directed cycle

File: doc/prop.xml [hunk 37]
```diff
@@ -852,3 +852,42 @@ true
...
+<#GAPDoc Label="IsHamiltonianDigraph">
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If digraph is Hamiltonian, this property returns
+    <K>true</K>, and <K>false</K> if it is not. <P/>
+
+    A digraph is <E>Hamiltonian</E> if it has a directed circuit visiting
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
```
Comment (wilfwilson) [inline doc/prop.xml]: Change it to `if it has a directed cycle of length <C>n</C>.`

File: doc/prop.xml [hunk 38]
```diff
@@ -852,3 +852,42 @@ true
...
 
+<#GAPDoc Label="IsHamiltonianDigraph">
+<ManSection>
+  <Prop Name="IsHamiltonianDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    If digraph is Hamiltonian, this property returns
+    <K>true</K>, and <K>false</K> if it is not. <P/>
+
+    A digraph is <E>Hamiltonian</E> if it has a directed circuit visiting
```
Comment (wilfwilson) [inline doc/prop.xml]: Change it to `A digraph with <C>n</C> vertices is...`

File: doc/prop.xml [hunk 39]
```diff
@@ -852,3 +852,42 @@ true
...
+    each vertex exactly once. See Section <Ref Subsect="Definitions"
+     Style="Number" /> for the definition of a directed circuit.
+    Note the empty digraphs on 0 and 1 vertices are considered
+    Hamiltonian.<P/>
+
+    The method used in this operation has the worst case complexity as
+    DigraphMonomorphism. See <Ref Oper="DigraphMonomorphism"/> for information
```
Comment (wilfwilson) [inline doc/prop.xml]: Again, just have `...complexity as <Ref Oper="DigraphMonomorphism"/>.`

File: gap/attr.gd [hunk 40]
```diff
@@ -69,6 +69,7 @@ DeclareOperation("DIGRAPHS_MaximalSymmetricSubdigraph", [IsDigraph, IsBool]);
 
 DeclareAttribute("UndirectedSpanningTree", IsDigraph);
 DeclareAttribute("UndirectedSpanningForest", IsDigraph);
+DeclareAttribute("HamiltonianPath", [IsDigraph]);
```
Comment (wilfwilson) [inline gap/attr.gd]: Get rid of the square brackets here.

**Comment by wilfwilson** [comment]
You should check why this is happening:
```
gap> HamiltonianPath(CycleDigraph(512));
fail
```
It works fine for 511 and 513.

**Comment by wilfwilson** [comment]
@le27 what do you think the status of this is? It seems like you've addressed all of the comments that I've made so far. It seems like I've only reviewed the documentation so far, however, so if you think it's ready then I'll now review the code.

**Comment by le27** [comment]
As far as I can tell the code should be ready as it is now. Please check it for problems as it is.

**Comment by wilfwilson** [comment]
Will do!

File: gap/prop.gi [hunk 41]
```diff
@@ -567,3 +567,118 @@ function(gr)
...
+          fi;
+      od;
+      if not (checkMT or check41 or check42) then
+        break;
+      fi;
+    od;
+    if checkMT or check41 or check42 then
+      return true;
+    fi;
+
```
Comment (wilfwilson) [inline gap/prop.gi]: From this point on, I think you should replace the following by:
```
  fi;

  return HamiltonianPath(gr) <> fail;
end);
```

File: gap/prop.gi [hunk 42]
```diff
@@ -567,3 +567,118 @@ function(gr)
...
+  iter := IteratorOfPaths(gr, 1, 1);
+  n := n + 1;
+  while not IsDoneIterator(iter) do
+    if Length(NextIterator(iter)[1]) = n then
+      return true;
+    fi;
+  od;
+  return false;
+end);
```
Comment (wilfwilson) [inline gap/prop.gi]: After this, you should add the following method:
```
InstallMethod(IsHamiltonianDigraph, "for a digraph with hamiltonian path",
[IsDigraph and HasHamiltonianPath],
x -> HamiltonianPath(x) <> fail);
```
Because, if we've already calculated a path, we already know the answer.

Comment (wilfwilson) [reply gap/prop.gi]: And also, add a test for this case.

**Comment by wilfwilson** [comment]
Thanks @le27. Although @james-d-mitchell hasn't clicked "Approve" to the changes that he requested, it looks like you've addressed his concerns, so I'm happy to merge.
